logstash-core/src/main/java/org/logstash/plugins/pipeline/PipelineBus.java

```diff
@@ -17,31 +17,41 @@
  * under the License.
  */
 
-
 package org.logstash.plugins.pipeline;
 
 import com.google.common.annotations.VisibleForTesting;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.logstash.RubyUtil;
 import org.logstash.ext.JrubyEventExtLibrary;
 
-import java.util.Arrays;
 import java.util.Collection;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.stream.Stream;
+import java.util.Optional;
+import java.util.Set;
 
 /**
- * This class is the communication bus for the `pipeline` inputs and outputs to talk to each other.
- * <p>
- * This class is threadsafe.
+ * A {@code PipelineBus} is the communication bus that allows {@link PipelineOutput}s to
+ * send events to {@link PipelineInput}s.
+ *
+ * @implSpec implementations must be threadsafe for all operations.
  */
-public class PipelineBus {
+public interface PipelineBus {
 
-    final ConcurrentHashMap<String, AddressState> addressStates = new ConcurrentHashMap<>();
-    final ConcurrentHashMap<PipelineOutput, ConcurrentHashMap<String, AddressState>> outputsToAddressStates = new ConcurrentHashMap<>();
-    volatile boolean blockOnUnlisten = false;
-    private static final Logger logger = LogManager.getLogger(PipelineBus.class);
+    Logger LOGGER = LogManager.getLogger(PipelineBus.class);
+
+    /**
+     * API-stable entry-point for creating a {@link PipelineBus}
+     * @return a new pipeline bus
+     */
+    static PipelineBus create() {
+        final String pipelineBusImplementation = System.getProperty("logstash.pipelinebus.implementation", "v1");
```

Comment:
I don't see a reason to default to v1 here. We should offer an escape hatch by allowing users to downgrade to v1, but v2 is faster and shouldn't break compatibility. My suggestion is to switch to v2 and already mark v1 as deprecated.

Reply:
I agree. I primarily implemented this as a separate implementation of the extracted interface so that I could prove to _myself_ that it had the same behaviour. I have confidence that it is no _less_ safe than the v1 current implementation, but since I had already done the extraction I don't see harm in leaving v1 as an escape hatch.